File: Irrgarten/src/irrgarten/Player.java
package irrgarten;
import java.util.ArrayList;

/**
 * The Player class represents a player in the game.
 */
public class Player extends LabyrinthCharacter {

    /*s
    
    * The maximum number of weapons a player can have.
    */
    public static final int MAX_WEAPONS = 2;

    /**
     * The maximum number of shields a player can have.
     */
    public static final int MAX_SHIELDS = 3;

    /**
     * The initial health of the player.
     */
    public static final int INITIAL_HEALTH = 10;

    /**
     * The number of consecutive hits a player can receive before losing.
     */
    public static final int HITS2LOSE = 3;
    
    /**
     * The player number.
     */
    private final char number;
    
    /**
     * The number of consecutive hits the player has received.
     */
    private int consecutiveHits;

    /**
     * The player's weapons and shields.
     */
    private ArrayList<Weapon> weapons;
    private ArrayList<Shield> shields;

    private ShieldCardDeck shieldCardDeck;
    private WeaponCardDeck weaponCardDeck;
    /**
     * Parametrized constructor to create a Player object.
     * @param number The player number.
     * @param intelligence The intelligence level of the player.
     * @param strength The strength level of the player.
     */
    public Player(char number, float intelligence, float strength) {
        super("Player " + (char) number, intelligence, strength, (float) Player.INITIAL_HEALTH);
        this.number = number;
        this.weapons = new ArrayList<Weapon>(0);
        this.shields = new ArrayList<Shield> (0);
        this.shieldCardDeck = new ShieldCardDeck();
        this.weaponCardDeck = new WeaponCardDeck();
        this.resetHits();
    }

    public Player(Player other){
        super(other);
        this.number = other.number;
        this.weapons = other.weapons;
        this.shields = other.shields;
        this.shieldCardDeck = other.shieldCardDeck;
        this.weaponCardDeck = other.weaponCardDeck;
        this.consecutiveHits = other.consecutiveHits;
    }
    /**
     * Resurrects the player by resetting their attributes.
     */
    public void resurrect() {
        this.weapons.clear();
        this.shields.clear();
        this.setHealth(Player.INITIAL_HEALTH);
        this.resetHits();
    }
    
    /**
     * Gets the player number.
     * @return The player number.
     */
    public char getNumber(){
        return this.number;
    }

    /**
     * Moves the player in the specified direction.
     * @param direction The direction to move.
     * @param validMoves The list of valid moves.
     * @return The direction the player moved to.
     */
    public Directions move(Directions direction, ArrayList<Directions> validMoves) {
        int size = validMoves.size();
        boolean contained = validMoves.contains(direction);
        if(size > 0  && !contained){
            Directions firstElement = validMoves.get(0);
            return firstElement;
        }
        else
            return direction;
    }

    /**
     * Calculates the attack power of the player.
     * @return The attack power.
     */
    @Override
    public float attack(){
        return this.getStrength() + this.sumWeapons();
    }
    
    /**
     * Defends against an attack.
     * @param receivedAttack The attack power received.
     * @return true if the player successfully defended, false otherwise.
     */
    @Override
    public boolean defend(float receivedAttack){
        return this.manageHit(receivedAttack);
    }
    
    /**
     * Receives a reward for defeating a monster.
     * The reward consists of new weapons, shields, and health.
     * The number of weapons and shields and the health reward received is determined by the dice.
     * 
     */
    public void receiveReward(){	
        int wReward = Dice.weaponsReward();
        int sReward = Dice.shieldsReward();
        for (int i= 0; i < wReward; i++){
            Weapon wNew = this.newWeapon();
            this.receiveWeapon(wNew);
        }
        
        for (int i= 0; i < sReward; i++){
            Shield sNew = this.newShield();
            this.receiveShield(sNew);
        }
        
        int extraHealth = Dice.healthReward();
        float newHealth = this.getHealth() + extraHealth;
        this.setHealth(newHealth);
    }
    
    /**
     * Returns a string representation of the player.
     * @return The string representation of the player.
     */
    @Override
    public String toString(){
        String ret = "P[" + super.toString()
                    + ", ConsecHits: " + Integer.toString(this.consecutiveHits) 
                    + "]\n\tWeapons: ";
	    // Show the weapons
	    for (Weapon aWeapon : this.weapons) {
	        ret +=  "\n\t" + aWeapon.toString();
	    }
	    ret += "\n\tShields: ";
	    for (Shield aShield : this.shields) {
	        ret += "\n\t" + aShield.toString();
	    }
            ret += "\n";
        return ret;
    }
    
    /**
     * Receives a weapon.
     * @param w The weapon to receive.
     */
    private void receiveWeapon(Weapon w){
        
        this.weapons.removeIf(wi -> wi.discard());
        int size = this.weapons.size();
        if(size < Player.MAX_WEAPONS)
            this.weapons.add(w);
    }

    /**
     * Receives a shield.
     * @param s The shield to receive.
     */
    private void receiveShield(Shield s){
        
        this.shields.removeIf(si -> si.discard());
        int size = this.shields.size();
        if(size < Player.MAX_SHIELDS)
            this.shields.add(s);

    }
    
    /**
     * Creates a new weapon.
     * @return The new weapon.
     */
    private Weapon newWeapon(){
        return this.weaponCardDeck.nextCard();
    }

    /**
     * Creates a new shield.
     * @return The new shield.
     */
    private Shield newShield(){
        return this.shieldCardDeck.nextCard();
    }
    
    /**
     * Calculates the sum of the weapons' attack power.
     * @return The sum of the weapons' attack power.
     */
    protected float sumWeapons(){
        float sum = 0; 
        for (Weapon aWeapon : this.weapons){
            sum += aWeapon.attack();
        }
        return sum;
    }
    
    /**
     * Calculates the sum of the shields' protection.
     * @return The sum of the shields' protection.
     */
    protected float sumShields(){
        float sum = 0;
        for (Shield aShield : this.shields){
            sum += aShield.protect();
        }
        return sum;
    }
    
    /**
     * Calculates the defensive energy of the player.
     * @return The defensive energy.
     */
    protected float defensiveEnergy(){
        return this.sumShields() + this.getIntelligence();
    }

    /**
     * Manages the hit received by the player.
     * @param receivedAttack The attack power received.
     * @return true if the player lost, false otherwise.
     */
    private boolean manageHit(float receivedAttack){
        float defense = this.defensiveEnergy();
        if(defense < receivedAttack){
            this.gotWounded();
            this.incConsecutiveHits();
        }
        else
            this.resetHits();
        
        boolean lose;
        if( (this.consecutiveHits == Player.HITS2LOSE) || dead()){
            this.resetHits();
            lose = true;
        }
        else
            lose = false;
        
        return lose;
    }
    
    /**
     * Resets the consecutive hits counter.
     */
    private void resetHits(){
        this.consecutiveHits = 0;
    }

    /**
     * Increases the consecutive hits counter.
     */
    private void incConsecutiveHits(){
        this.consecutiveHits++;
    }
}

File: Irrgarten_ruby/player.rb
#encoding:utf-8
require_relative 'dice'
require_relative 'weapon'
require_relative 'shield'
require_relative 'labyrinth_character'
module Irrgarten
    # The Player class represents a player in the Irrgarten game.
    class Player < LabyrinthCharacter
        # The maximum number of weapons a player can have.
        @@MAX_WEAPONS = 2

        # The maximum number of shields a player can have.
        @@MAX_SHIELDS = 3

        # The initial health of a player.
        @@INITIAL_HEALTH = 10

        # The number of consecutive hits a player can take before losing.
        @@HITS2LOSE = 3

        # Initializes a new instance of the Player class.
        #
        # @param number [Integer] The player number.
        # @param intelligence [Integer] The player's intelligence level.
        # @param strength [Integer] The player's strength level.
        def initialize(number, intelligence, strength)
            @number = number
            super("Player #{@number}", intelligence, strength, @@INITIAL_HEALTH)
            self.resurrect
        end

        # Resurrects the player, resetting their state.
        def resurrect
            @weapons = Array.new
            @shields = Array.new
            @health = @@INITIAL_HEALTH
            self.reset_hits
        end

        # Copy constructor
        # @param other Player to copy
        def copy(other)
            super(other)
            @consecutive_hits = other.consecutive_hits
            @number = other.number
            @weapons = other.weapons
            @shields = other.shields
        end
        attr_reader :row, :col, :number, :consecutive_hits,  :weapons, :shields

        # Moves the player in a given direction.
        # If the direction is not valid, the player will move in the first valid direction.
        # If there are no valid directions, the player will stay in the same position.
        # @param direction [Symbol] The direction to move.
        # @param valid_moves [Array<Symbol>] The valid directions to move.
        # @return [Symbol] The direction the player moved.
        def move(direction, valid_moves)
            size = valid_moves.size
            contained = valid_moves.include?(direction)
            if size > 0 && !contained
                valid_moves[0]
            else
                direction
            end
        end

        # Calculates the player's attack power.
        #
        # @return [Integer] The player's attack power.
        def attack
            @strength + sum_weapons
        end

        # Defends against an attack.
        #
        # @param received_attack [Integer] The attack power received.
        def defend(received_attack)
            manage_hit(received_attack)
        end

        # Receives a reward for winning a combat.
        # The reward consists of new weapons, shields, and health points.
        # The number of health points, weapons and shields is determined by the Dice class.
        def receive_reward
            w_reward = Dice.weapons_reward
            s_reward = Dice.shields_reward
            w_reward.times do
                wnew = self.new_weapon
                self.receive_weapon(wnew)
            end
            s_reward.times do
                snew = self.new_shield
                self.receive_shield(snew)
            end
            extra_health = Dice.health_reward
            @health += extra_health
        end

        # Returns a string representation of the player's state.
        #
        # @return [String] The player's state.
        def to_s
            ret = "P[#{super}, ConsecHits: #{@consecutive_hits.to_s}] \n\tWeapons: "
            @weapons.each { |a_weapon| ret += "\n\t#{a_weapon.to_s}" }
            ret += "\n\tShields:"
            @shields.each { |a_shield| ret += "\n\t#{a_shield.to_s}" }
            ret
        end

        private

        # Receives a weapon.
        #
        # @param w [Weapon] The weapon to receive.
        def receive_weapon(w)
            @weapons.delete_if { |wi| wi.discard }
            size = @weapons.length
            if size < @@MAX_WEAPONS
                @weapons << w
            end
        end

        # Receives a shield.
        #
        # @param s [Shield] The shield to receive.
        def receive_shield(s)
            @shields.delete_if { |si| si.discard }
            size = @shields.length
            if size < @@MAX_SHIELDS
                @shields << s
            end
        end

        # Creates a new weapon.
        #
        # @return [Weapon] The new weapon.
        def new_weapon
            r_power = Dice.weapon_power
            r_uses = Dice.uses_left
            Weapon.new(r_power, r_uses)
        end

        # Creates a new shield.
        #
        # @return [Shield] The new shield.
        def new_shield
            r_protection = Dice.shield_power
            r_uses = Dice.uses_left
            Shield.new(r_protection, r_uses)
        end

        protected

        # Calculates the total attack power of all weapons.
        #
        # @return [Integer] The total attack power.
        def sum_weapons
            sum = 0
            @weapons.each { |a_weapon| sum += a_weapon.attack }
            sum
        end

        # Calculates the total protection power of all shields.
        #
        # @return [Integer] The total protection power.
        def sum_shields
            sum = 0
            @shields.each { |a_shield| sum += a_shield.protect }
            sum
        end

        # Calculates the defensive energy of the player.
        #
        # @return [Integer] The defensive energy.
        def defensive_energy
            self.sum_shields + @intelligence
        end

        private

        # Manages the player's state when hit.
        # If the player is hit, the consecutive hits counter is increased.
        # If the player is not hit, the counter is reset.
        # If the player has received the maximum number of consecutive hits, the player loses.
        # @param received_attack [Integer] The attack power received.
        # @return [Boolean] True if the player has lost, false otherwise.
        def manage_hit(received_attack)
            defense = self.defensive_energy
            if defense < received_attack
                self.got_wounded
                self.inc_consecutive_hits
            else
                self.reset_hits
            end
            if ((@consecutive_hits == @@HITS2LOSE) || self.dead)
                self.reset_hits
                lose = true
            else
                lose = false
            end
            lose
        end

        # Resets the consecutive hits counter.
        def reset_hits
            @consecutive_hits = 0
        end<

        # Increases the consecutive hits counter.
        def inc_consecutive_hits
            @consecutive_hits += 1
        end
    end
end
